tools/AzureRM.BootStrapper/AzureRM.Bootstrapper.psm1

```diff
@@ -66,38 +146,478 @@ function Get-AzProfile
+  }
+  else
+  {
+    Install-Module $Module -RequiredVersion $version -Repository $BootStrapRepo -scope $Scope -AllowClobber
+  }
+}
+
+# Function to remove Previous profile map
+function Remove-ProfileMapFile
+{
+  [CmdletBinding()]
+	param([string]$ProfileMapPath)
+
+  if (Test-Path -Path $ProfileMapPath)
+  {
+    RemoveWithRetry -Path $ProfileMapPath -Force
+  }
+}
+
+# Remove-Item command with Retry
+function RemoveWithRetry
+{
+  param ([string]$Path)
+
+  $retries = 3
+  $secondsDelay = 2
+  $retrycount = 0
+  $completedSuccessfully = $false
+
+  while (-not $completedSuccessfully) 
+  {
+    try 
+    {
+      Remove-Item @PSBoundParameters -ErrorAction Stop
+      $completedSuccessfully = $true
+    } 
+    catch 
+    {
+      if ($retrycount -ge $retries) 
+      {
+        throw
+      } 
+      else 
+      {
+        Start-Sleep $secondsDelay
+        $retrycount++
+      }
+    }
+  }
+}
+
+# Get profiles installed associated with the module version
+function Test-ProfilesInstalled
+{
+  param([String]$Module, [String]$Profile, [PSObject]$PMap, [hashtable]$AllProfilesInstalled)
+
+  # Profiles associated with the particular module version - installed?
+  $profilesAssociated = @()
+  $versionList = $PMap.$Profile.$Module
+  foreach ($version in $versionList)
+  {
+    foreach ($profileInAllProfiles in $AllProfilesInstalled[$Module + $version])
+    {
+      $profilesAssociated += $profileInAllProfiles
+    }
+  }
+  return $profilesAssociated
+}
+
+# Function to uninstall module
+function Uninstall-ModuleHelper
+{
+  [CmdletBinding(SupportsShouldProcess = $true)] 
+  param([PSObject]$PMap, [String]$Module, [System.Version]$version, [hashtable]$AllProfilesInstalled)
+  DynamicParam
+  {
+    $params = New-Object -Type System.Management.Automation.RuntimeDefinedParameterDictionary
+    Add-ProfileParam $params
+    Add-RemoveParam $params
+    return $params
+  }
+
+  PROCESS
+  {
+    $Profile = $PSBoundParameters.Profile
+    $Remove = $PSBoundParameters.RemovePreviousVersions
+
+    Do
+    {
+      $moduleInstalled = Get-Module -Name $Module -ListAvailable | Where-Object { $_.Version -eq $version} 
+      if ($PSCmdlet.ShouldProcess("$module version $version", "Remove module")) 
+      {
+        if (($moduleInstalled -ne $null) -and ($Remove.IsPresent -or $PSCmdlet.ShouldContinue("Remove module $Module version $version", "Removing Modules for profile $Profile"))) 
+        {
+          Remove-Module -Name $module -Force -ErrorAction "SilentlyContinue"
+          try 
+          {
+            Uninstall-Module -Name $module -RequiredVersion $version -Force -ErrorAction Stop
+          }
+          catch
+          {
+            Write-Warning $_.Exception.Message
+            break
+          }
+        }
+        else {
+          break
+        }
+      }
+      else {
+        break
+      }
+    }
+    While($moduleInstalled -ne $null);        
```

Comment:
We haven't added retry for Uninstall-AzureRmProfile. The user will get a warning if it throws an error and user has to retry.